pkg/controller/jaeger/jaeger_controller.go

```diff
@@ -0,0 +1,184 @@
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileJaeger) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.WithFields(log.Fields{
+		"namespace": request.Namespace,
+		"name":      request.Name,
+	}).Print("Reconciling Jaeger")
+
+	// Fetch the Jaeger instance
+	instance := &v1alpha1.Jaeger{}
+	err := r.client.Get(context.Background(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// workaround for https://github.com/kubernetes-sigs/controller-runtime/issues/202
+	// see also: https://github.com/kubernetes-sigs/controller-runtime/pull/212
+	// once there's a version incorporating the PR above, the manual setting of the GKV can be removed
+	instance.APIVersion = fmt.Sprintf("%s/%s", v1alpha1.SchemeGroupVersion.Group, v1alpha1.SchemeGroupVersion.Version)
+	instance.Kind = "Jaeger"
+
+	ctrl := NewController(context.Background(), instance)
+
+	// wait for all the dependencies to succeed
+	if err := r.handleDependencies(ctrl); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	created, err := r.handleCreate(ctrl)
+	if err != nil {
+		log.WithField("instance", instance).WithError(err).Error("failed to create")
+		return reconcile.Result{}, err
+	}
+
+	if created {
+		log.WithField("name", instance.Name).Info("Configured Jaeger instance")
+	}
+
+	if err := r.handleUpdate(ctrl); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// we store back the changed CR, so that what is stored reflects what is being used
+	if err := r.client.Update(context.Background(), instance); err != nil {
+		log.Errorf("failed to update %v", instance)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileJaeger) handleCreate(ctrl Controller) (bool, error) {
+	objs := ctrl.Create()
+	created := false
+	for _, obj := range objs {
+		err := r.client.Create(context.Background(), obj)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			log.Errorf("failed to create %v", obj)
+			return false, err
+		}
+
+		if err == nil {
+			created = true
+		}
+	}
+
+	return created, nil
+}
+
+func (r *ReconcileJaeger) handleUpdate(ctrl Controller) error {
+	objs := ctrl.Update()
+	for _, obj := range objs {
+		log.Debugf("Updating %v", obj)
+		if err := r.client.Update(context.Background(), obj); err != nil {
+			log.Errorf("failed to update %v", obj)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcileJaeger) handleDependencies(ctrl Controller) error {
+	for _, dep := range ctrl.Dependencies() {
```

Comment:
Can the dependencies be launched concurrently - rather than waiting on each one in turn?

Reply:
Yes, I believe so. Would you open an issue for that? This is an existing code that just got moved here.

Reply:
Done #117